### PR TITLE
Frontend/license detail UI

### LINKED ISSUE #67: frontend licensee detail view
### Acceptance Criteria
- Personal information
- License information
- Adverse action simple view

### PR DESCRIPTION
**Pull request by jsandoval81**
### Requirements
- [x] Merge #100  _**(CSG)**_
    - _#100 has API updates on which this frontend PR is dependent_
- [x] Merge latest upstream `development` into this PR _**(IA)**_
- [x] Confirm merged functionality & publish PR _**(IA)**_

### Description List
- Add licensee detail UI page
- Update licensee list page to link to the new licensee detail page
- Update data layer for API endpoint changes
- Update Dev deployment config for API service changes

Closes #67

### CODE REVIEW

**Comment by jlkravitz** [comment]
@jsandoval81 We just merged #100 - I'll hold off on reviewing this one until you've merged the development branch into this one and resolved conflicts (unless you tell me otherwise).

**Comment by jsandoval81** [comment]
@jlkravitz I just got the latest `development` merged here & published this PR. Should be ready for your review.

**Comment by jlkravitz** [comment]
- [ ] review the pull request to get oriented
	- [ ] read the description of the pull request, which should summarize the changes made
	- [ ] read through every task on the Scrum board that's encompassed by this pull request
	- [ ] read the description of the commits that comprise the pull request
- [ ] stand up the site locally
	- [ ] test all functionality in all major browsers, emphasizing the functionality that this pull request addresses
		- [ ] for public-facing functionality, test in browsers consistent with [public browser use data](https://analytics.usa.gov/)
		- [ ] test in Mobile Safari and Mobile Chrome
	- [ ] use an automated audit tool for code quality and practices (recommended: [Chrome DevTools](https://developers.google.com/web/tools/chrome-devtools/), aka [Lighthouse](https://chrome.google.com/webstore/detail/lighthouse/blipmdconlkpinefehnmjammfjpmpbjk))
		- [ ] look at efficiency of page loads, asset sizes, HTTP connection management, etc.
	- [ ] review for accessibility
		- [ ] use an automated audit tool, such as Chrome Audit or aXe
		- [ ] navigate site only with the keyboard
		- [ ] use VoiceOver or Narrator to navigate the site with audio only, with the display turned off
		- [ ] manually test anything that pa11y cannot test automatically (e.g., contrast of text over images)
- [ ] review static code analysis results
- [ ] run a security audit of dependencies (e.g. `npm audit`) to ensure that there are no vulnerabilities that will be deployed to production (as opposed to vulnerabilities that only have an impact on the development environment)
- [ ] examine OWASP ZAP output to ensure that any errors are known to be false positives or have been previously declared to be acceptable
- [ ] skim all new code, in the context of existing code, looking for problems (knowing that the vast majority of new code will be covered by tests)
- [ ] review all tests
	- [ ] methodically review all new tests for correctness, quality of naming
	- [ ] look at code coverage of tests
	- [ ] determine what code isn’t tested, review that rigorously
- [ ] review documentation to ensure that it matches changes
- [ ] provide comments on the pull request on GitHub, as necessary
	- [ ] for comments that are specific to a particular line of code, comment on those specific lines
	- [ ] for comments that are more general, attach the comment to a random line in `README.md` (as opposed to commenting on the pull request itself), to be able to use GitHub's ability to thread discussions on those comments
- [ ] for each feature-level bug (i.e., it’s working as designed, but designed wrong), open a new issue and put it in the backlog

File: webroot/src/components/Licensee/LicenseeRow/LicenseeRow.vue
Comment (jlkravitz) [inline webroot/src/components/Licensee/LicenseeRow/LicenseeRow.vue:13]: Thanks for including keyboard navigation!

**Review by jlkravitz — CHANGES_REQUESTED** [review]
This looks great! Several (mostly nit-picky) comments, and a few questions. Will stand up the site and try out the feature shortly.

File: webroot/src/components/Licensee/LicenseeList/LicenseeList.ts
Comment (jsandoval81) [reply webroot/src/components/Licensee/LicenseeList/LicenseeList.ts:161]: It's bundled with the comment a few lines up about temporary limited server support. I can add a separate comment line here so it's clearer.

File: webroot/src/models/Licensee/Licensee.model.spec.ts
Comment (jsandoval81) [reply webroot/src/models/Licensee/Licensee.model.spec.ts:173]: Yep, exactly.

File: webroot/src/models/Licensee/Licensee.model.ts
Comment (jsandoval81) [reply webroot/src/models/Licensee/Licensee.model.ts:122]: I think it's right at the limit. Our linter rules currently have a max of 120.

File: webroot/src/network/licenseApi/data.api.ts
Comment (jsandoval81) [reply webroot/src/network/licenseApi/data.api.ts:99]: > To confirm my understanding: Does the value of licenseeId determine if this is a detail or list view request?

Yep. I don't know if it's been settled yet if the server will use this same prop for search (upcoming); as in whether it could eventually accept partial licensee ids that could potentially return many records.

> Could we reduce nesting by one layer by just returning in the above if statement? If that goes against your style guide, ignore me.

We prefer not to have early `return`s.

File: webroot/src/network/licenseApi/data.api.ts
Comment (jsandoval81) [reply webroot/src/network/licenseApi/data.api.ts:149]: For now that's just the version of the API that's unauthenticated and returns mock records. Once we have the backend further along, I think that path segment will become the actual API version (e.g. `/v1`).

File: webroot/src/network/mocks/mock.data.api.ts
Comment (jsandoval81) [reply webroot/src/network/mocks/mock.data.api.ts:88]: If the licensee record was found in the local mock data by ID, then the successful mock response is returned. Otherwise an error response is returned.

The `wait(500)` is just a small delay to make sure our loading UI is working.

File: webroot/src/network/stateApi/data.api.ts
Comment (jsandoval81) [reply webroot/src/network/stateApi/data.api.ts:76]: Same as above. I think the server just switched from `/v0` to `/mock` for its mock-style endpoints.

Comment (jsandoval81) [reply webroot/src/network/stateApi/data.api.ts:76]: Also, I should probably clarify: This file is hitting a real remote endpoint, which includes `/mock` in the URI. Alternatively, the `mock.data.api` in this frontend app is a local-only mock API that we switch to when (e.g.) the remote endpoints aren't ready yet, are having an outage, etc. So there are currently 2 mock APIs - 1 remote & 1 local. Once the remote API is working for real, the paths here in this file will get updated from `/mock` to ~`/v1`.

File: webroot/src/pages/LicensingDetail/LicensingDetail.vue
Comment (jsandoval81) [reply webroot/src/pages/LicensingDetail/LicensingDetail.vue:89]: Yep, it's in the designs as a placeholder for now. Good double check 👍

File: webroot/src/pages/LicensingDetail/LicensingDetail.vue
Comment (jsandoval81) [reply webroot/src/pages/LicensingDetail/LicensingDetail.vue:101]: Yep, it's in the designs as a placeholder for now. Good double check 👍

File: webroot/src/store/license/license.getters.ts
Comment (jsandoval81) [reply webroot/src/store/license/license.getters.ts:10]: We have a TODO to get types fully working with our Vuex stores. IIRC I think there were some non-trivial steps there. Would you suggest we try to make that a prioritized task for the next sprint? Or just get it in the backlog somewhere for now?

File: webroot/src/network/licenseApi/data.api.ts
Comment (jsandoval81) [reply webroot/src/network/licenseApi/data.api.ts:133]: We've generally left our server responses as any, and then handle the transformation into typed models in our model serializers. This has seemed to reduce friction at the network layer, but still allow us to use strongly-typed models in our store. Does this seem ok? Or would you prefer we use stronger typing in each of these cases?

File: webroot/src/pages/LicensingDetail/LicensingDetail.ts
Comment (jsandoval81) [reply webroot/src/pages/LicensingDetail/LicensingDetail.ts:49]: Fair question. The main reasoning is because `licenseeId` is reused on the following line. Generally we'll try cache vars that are used more than once, both for readability & also avoid the potential cost of traversing the parent object multiple times.

File: webroot/src/pages/LicensingDetail/LicensingDetail.ts
Comment (jsandoval81) [reply webroot/src/pages/LicensingDetail/LicensingDetail.ts:76]: I think that should work. We'll just be required to add the optional chaining operator, and possibly cast the fallback value:

```
get licenseeLicenses(): Array<License> {
    return this.licensee?.licenses || [] as Array<License>;
}
```

That's probably still in-bounds for readability.

File: webroot/src/pages/LicensingDetail/LicensingDetail.ts
Comment (jsandoval81) [reply webroot/src/pages/LicensingDetail/LicensingDetail.ts:65]: I think that should work. We'll just be required to add the optional chaining operator:

```
get licenseeNameDisplay(): string {
    return this.licensee?.nameDisplay() || '';
}
```

File: webroot/src/pages/LicensingDetail/LicensingDetail.ts
Comment (jsandoval81) [reply webroot/src/pages/LicensingDetail/LicensingDetail.ts:54]: This might be one that I'd prefer to leave closer to as-is. In general, our team is going to prefer readability / self-documenting over clever one-liners. Open to discussion if you feel strongly.

File: webroot/src/store/license/license.actions.ts
Comment (jsandoval81) [reply webroot/src/store/license/license.actions.ts:45]: I think this was just a separation of concerns. The "success" here is more in the context of a `200`-like response from the server. This action doesn't necessarily care about the content of `licensee`, just if it's not falsy then it commits it to the store. The UI component would ultimately be the place that decides what to display to the user ("not found", redirect, etc). Open to follow-up thoughts.

File: webroot/src/store/license/license.spec.ts
Comment (jsandoval81) [reply webroot/src/store/license/license.spec.ts:261]: It's just the common prop that each of our stores have. It'd be similar to calling it `records`. I think originally we wanted to express that they weren't just records in a server context, but serialized model instances that were ready to be used in components. Nothing too fancy 🙂

File: backend/compact-connect/README.md
Comment (jsandoval81) [reply backend/compact-connect/README.md:160]: We can create a backlog item to incorporate `yarn test:unit:coverage` into the automation.

`yarn audit:dependencies` is currently part of `yarn build` (and `yarn serve`). If you'd prefer it be its own step, we can add that to the backlog item.

Comment (jlkravitz) [reply backend/compact-connect/README.md:160]: Re: `yarn audit:dependencies`, missed that! That's fine.

Re: `yarn test:unit:coverage`, sounds good. Not high priority for now, but would be great to have within a sprint or two.

File: webroot/src/models/Licensee/Licensee.model.ts
Comment (jlkravitz) [reply webroot/src/models/Licensee/Licensee.model.ts:122]: Great!

File: webroot/src/network/licenseApi/data.api.ts
Comment (jlkravitz) [reply webroot/src/network/licenseApi/data.api.ts:99]: Got it. Thanks for the explanation. Good leaving as-is!

File: webroot/src/network/licenseApi/data.api.ts
Comment (jlkravitz) [reply webroot/src/network/licenseApi/data.api.ts:133]: That makes sense! I really just want to make sure use of `any` is generally limited and intentional. Good to continue that convention.

File: webroot/src/network/mocks/mock.data.api.ts
Comment (jlkravitz) [reply webroot/src/network/mocks/mock.data.api.ts:88]: Ah I totally missed that this was a mock of the data API (for testing, I assume). Thanks for the explanation!

Could we add a brief comment on the intent of `this.wait(500)`?

Comment (jlkravitz) [reply webroot/src/network/mocks/mock.data.api.ts:88]: You probably have good reason for the naming, but just curious: any reason to not name it something like `local.data.api.ts`? My first instinct was that this was for testing (once I noticed it was a "mock" file).

File: webroot/src/network/stateApi/data.api.ts
Comment (jlkravitz) [reply webroot/src/network/stateApi/data.api.ts:76]: Got it thanks!

File: webroot/src/pages/LicensingDetail/LicensingDetail.ts
Comment (jlkravitz) [reply webroot/src/pages/LicensingDetail/LicensingDetail.ts:49]: Fair enough. As long as we're consistent, this is good with me!

File: webroot/src/pages/LicensingDetail/LicensingDetail.ts
Comment (jlkravitz) [reply webroot/src/pages/LicensingDetail/LicensingDetail.ts:54]: That's fine with me. This was certainly the one where I thought to myself that it might be worth not making this a one-liner.

File: webroot/src/pages/LicensingDetail/LicensingDetail.ts
Comment (jlkravitz) [reply webroot/src/pages/LicensingDetail/LicensingDetail.ts:65]: Sounds good. I'll add that these are all suggestions, and just _my_ preference but it's generally not my business to change your teams' conventions. More important to me is consistency across the codebase!

File: webroot/src/pages/LicensingDetail/LicensingDetail.ts
Comment (jlkravitz) [reply webroot/src/pages/LicensingDetail/LicensingDetail.ts:76]: Sounds good! Is the cast required?

File: webroot/src/pages/LicensingDetail/LicensingDetail.vue
Comment (jlkravitz) [reply webroot/src/pages/LicensingDetail/LicensingDetail.vue:89]: Got it. Going forward, as we start to move towards PRs centered around user stories, I'd prefer fully-implemented code without filler data if possible (same goes for ToDo comments, commented-out code, etc).

In this case, if I understand correctly, that might mean leaving out the code for the adverse reactions until the backend is properly set up. @jusdino just FYA since this is relevant to backend code, too.

I'm fine keeping this as-is if that's easiest for now.

File: webroot/src/store/license/license.actions.ts
Comment (jlkravitz) [reply webroot/src/store/license/license.actions.ts:45]: Got it, makes sense to me! In that case, any reason to not just _always_ pass `licensee` to the store? (Presumably, the UI would handle this case, too.)

File: webroot/src/store/license/license.getters.ts
Comment (jlkravitz) [reply webroot/src/store/license/license.getters.ts:10]: Got it! I'd be curious to learn more why it's not as simple as changing the type here to, e.g., `string` (I'm sure it's not), but that can be a face-to-face conversation. Won't make you spell it out here :)

I'd like to spread out technical tasks, so let's not prioritize for next sprint. That said, I might suggest we include in the backlog item a task to build CI/CD stats on `any` usage, so we can make it a priority over the project's lifetime. (Open to other CI/CD concepts here, too.)

File: webroot/src/store/license/license.spec.ts
Comment (jlkravitz) [reply webroot/src/store/license/license.spec.ts:261]: Got it, thanks!

File: webroot/src/pages/LicensingDetail/LicensingDetail.vue
Comment (jsandoval81) [reply webroot/src/pages/LicensingDetail/LicensingDetail.vue:89]: @jlkravitz That makes sense. I think @jusdino & I are flexible on this topic. We'll mainly want to make sure to sync up with product / design / stakeholder folks to manage expectations & figma designs around their demos & user-research interactions.

File: webroot/src/store/license/license.actions.ts
Comment (jsandoval81) [reply webroot/src/store/license/license.actions.ts:45]: It's a fair question. Ultimately, I think here it just boils down to "cache invalidation" - to update or not update the local cache (store) in this case. If you'd prefer to force remove the licensee from the store in this case, then we could explore updating the mutation to handle that (or creating a new mutation).

File: webroot/src/store/license/license.getters.ts
Comment (jsandoval81) [reply webroot/src/store/license/license.getters.ts:10]: I created #111 to track this.

File: backend/compact-connect/README.md
Comment (jsandoval81) [reply backend/compact-connect/README.md:160]: I created #112 to track this.

File: webroot/src/components/Licensee/LicenseeList/LicenseeList.ts
Comment (jsandoval81) [reply webroot/src/components/Licensee/LicenseeList/LicenseeList.ts:162]: Added a comment line above these for clarity.

File: webroot/src/network/licenseApi/data.api.ts
Comment (jsandoval81) [reply webroot/src/network/licenseApi/data.api.ts]: Updated this to be typed.

File: webroot/src/network/mocks/mock.data.api.ts
Comment (jsandoval81) [reply webroot/src/network/mocks/mock.data.api.ts:88]: We have some other scenarios where we have actual local APIs (where a backend stack is running on our dev machines). It's actually more unusual for us to have "mock" endpoints in the backend - and I think there is an upcoming ticket to remove those. So this frontend mock API will be the only mock API soon.

Also, I expanded the comment up at the definition of `wait` to give a clearer example of how it's used.

File: webroot/src/pages/LicensingDetail/LicensingDetail.ts
Comment (jsandoval81) [reply webroot/src/pages/LicensingDetail/LicensingDetail.ts]: Updated as we discussed. I wasn't 100% sure if cast would be needed - turns out it wasn't!

File: webroot/src/pages/LicensingDetail/LicensingDetail.ts
Comment (jsandoval81) [reply webroot/src/pages/LicensingDetail/LicensingDetail.ts]: Updated as we discussed.

File: webroot/src/store/license/license.spec.ts
Comment (jsandoval81) [reply webroot/src/store/license/license.spec.ts]: Fixed.

File: webroot/src/pages/LicensingDetail/LicensingDetail.vue
Comment (jlkravitz) [reply webroot/src/pages/LicensingDetail/LicensingDetail.vue:89]: Makes complete sense. And one clarification after chatting with @isabeleliassen and Waldo - most important to me is that the code from each _sprint_ is complete and operational, and implements specific user stories. I'm fine if front- and back- end code is broken up into two separate PRs, e.g.

File: webroot/src/store/license/license.actions.ts
Comment (jlkravitz) [reply webroot/src/store/license/license.actions.ts:45]: Helpful, thanks. Fine as-is, but perhaps add a quick note (e.g., if accurate, "only invalidate the local cache if we actually received something")?

Comment (jsandoval81) [reply webroot/src/store/license/license.actions.ts:45]: Done.

**Comment by jlkravitz** [comment]
A few notes from the rest of the checklist. Please address (as text response or in code) or put in the backlog, up to you. (Though going forward we should try to get down to zero medium / high-risk ZAP errors.)

## Manual Testing
* "Required field" shows up when I click "Select file", but I would only expect the error message to appear if I don't select a file from the file finder, or if I click submit without having tried to upload one.

## Lighthouse Report
* Seems that the "mobile network" simulation is so slow that the server returns `index.html`, which says "We're sorry but ia-vuejs-frontend doesn't work properly without JavaScript enabled. Please enable it to continue." Could we increase the timeout so that this doesn't happen? (Or prevent this from happening, however that may be done.)

## ZAP
Two medium-risk errors surfaced by OWASP ZAP below.

<details>
<summary>
Content Security Policy (CSP) Header Not Set (46)
</summary>
Content Security Policy (CSP) is an added layer of security that helps to detect and mitigate certain types of attacks, including Cross Site Scripting (XSS) and data injection attacks. These attacks are used for everything from data theft to site defacement or distribution of malware. CSP provides a set of standard HTTP headers that allow website owners to declare approved sources of content that browsers should be allowed to load on that page — covered types are JavaScript, CSS, HTML frames, fonts, images and embeddable objects such as Java applets, ActiveX, audio and video files.
</details>

<details>
<summary>
Missing Anti-clickjacking Header (46)
</summary>
The response does not include either Content-Security-Policy with 'frame-ancestors' directive or X-Frame-Options to protect against 'ClickJacking' attacks.
</details>

**Comment by jsandoval81** [comment]
@jlkravitz 

> In Chrome, but not Safari, "Required field" shows up when I click "Select file", but I would only expect the error message to appear if I don't select a file from the file finder, or if I click submit without having tried to upload one.

I created #113 to track this

> Lighthouse Report

I wasn't able to re-produce that result. Granted, LH / PSI heavily relies on the network speed of the client device. I opened #114 to track.

> ZAP

ZAP / CSP items are coming in a PR as soon as this one is merged (#64).

**Comment by jlkravitz** [comment]
@isabeleliassen This is good to merge!

**Review by jlkravitz — APPROVED** [review]

**Pull request merged**
